src/Haskell/Ide/Engine/Plugin/HsImport.hs

```diff
@@ -391,6 +399,40 @@ codeActionProvider plId docId _ context = do
         typeSig <- S.tailMay parts
         S.headMay typeSig
 
+      -- | Given a term 'data Sum f g a', extract the name of the type 'Sum'.
+      --
+      -- Can be used to obtain the Type name.
+      -- Can also be used to check whether the given term is from the
+      -- Type level or Value level.
+      -- Constructors and Types are sometimes indistinguishable,
+      -- e.g. querying for 'Sum', possible results include:
+      --
+      -- * "data Sum f g a"
+      -- * "Sum :: a -> Sum a"
+      --
+      -- Depending on whether we are looking for a Type or Value,
+      -- we want to generate different Code Actions.
+      -- Important to generate the correct Code Actions.
+      --
+      -- >>> typeDeclaration "data Sum f g a"
```

Comment:
Is this a doctest? Might be nice to turn it on in CI in some form

Reply:
Also we should probably stick in a test case somewhere

Reply:
It is intended as an usage example, but I think it is a doctest, too!
HsImport is vastly untested, unfortunately and quite a huge plugin. We need a way to extend these.